internal/controller/kubecertagent/deployer.go

```diff
@@ -0,0 +1,273 @@
+	for _, pod := range pods {
+		if pod.Status.Phase == corev1.PodRunning && (result == nil || newer(pod, result)) {
+			result = pod
+		}
+	}
+	return result
+}
+
+func int32ptr(i int32) *int32 { return &i }
+func int64ptr(i int64) *int64 { return &i }
+func boolptr(b bool) *bool    { return &b }
+
+func newAgentDeployment(c *AgentPodConfig, controllerManagerPod *corev1.Pod) *appsv1.Deployment {
+	var volumeMounts []corev1.VolumeMount
+	if len(controllerManagerPod.Spec.Containers) > 0 {
+		volumeMounts = controllerManagerPod.Spec.Containers[0].VolumeMounts
+	}
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      strings.TrimSuffix(c.NamePrefix, "-"), // TODO
+			Namespace: c.Namespace,
+			Labels:    c.AdditionalLabels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: int32ptr(1),
+			Selector: metav1.SetAsLabelSelector(c.Labels()),
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: c.Labels(),
+				},
+				Spec: corev1.PodSpec{
+					TerminationGracePeriodSeconds: int64ptr(0),
+					ImagePullSecrets:              c.ContainerImagePullSecretReferences(),
+					Containers: []corev1.Container{
+						{
+							Name:            "sleeper",
+							Image:           c.ContainerImage,
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command:         []string{"/bin/sleep", "infinity"},
+							VolumeMounts:    volumeMounts,
+							Env: []corev1.EnvVar{
+								{Name: "CERT_PATH", Value: getContainerArgByName(controllerManagerPod, "cluster-signing-cert-file", "/etc/kubernetes/ca/ca.pem")},
+								{Name: "KEY_PATH", Value: getContainerArgByName(controllerManagerPod, "cluster-signing-key-file", "/etc/kubernetes/ca/ca.key")},
+							},
+							Resources: corev1.ResourceRequirements{
+								Limits: corev1.ResourceList{
+									corev1.ResourceMemory: resource.MustParse("16Mi"),
+									corev1.ResourceCPU:    resource.MustParse("10m"),
+								},
+								Requests: corev1.ResourceList{
+									corev1.ResourceMemory: resource.MustParse("16Mi"),
+									corev1.ResourceCPU:    resource.MustParse("10m"),
+								},
+							},
+						},
+					},
+					Volumes:                      controllerManagerPod.Spec.Volumes,
+					RestartPolicy:                corev1.RestartPolicyAlways,
+					NodeSelector:                 controllerManagerPod.Spec.NodeSelector,
+					AutomountServiceAccountToken: boolptr(false),
+					NodeName:                     controllerManagerPod.Spec.NodeName,
+					Tolerations:                  controllerManagerPod.Spec.Tolerations,
+					// We need to run the agent pod as root since the file permissions
+					// on the cluster keypair usually restricts access to only root.
+					SecurityContext: &corev1.PodSecurityContext{
+						RunAsUser:  int64ptr(0),
+						RunAsGroup: int64ptr(0),
+					},
+				},
+			},
+			MinReadySeconds: 10,
+		},
+	}
+}
+
+func mergeObjectMeta(existing metav1.ObjectMeta, expected metav1.ObjectMeta) metav1.ObjectMeta {
```

Comment:
:thinking: I think there is a helper somewhere for this.

Reply:
I found the OpenShift helper but not anything in core kubernetes: https://github.com/openshift/library-go/blob/7a65fdb398e28782ee1650959a5e0419121e97ae/pkg/operator/resource/resourcemerge/object_merger.go#L12